Lib/test/test_importlib/import_/test_path.py

```diff
@@ -153,6 +154,25 @@ def test_deleted_cwd(self):
             # Do not want FileNotFoundError raised.
             self.assertIsNone(self.machinery.PathFinder.find_spec('whatever'))
 
+    @os_helper.skip_unless_working_chmod
+    def test_permission_error_cwd(self):
+        # gh-115911
+        with (
+            os_helper.temp_dir() as new_dir,
+            os_helper.save_mode(new_dir),
+            os_helper.save_cwd(),
+            util.import_state(path=['']),
+        ):
+            os.chdir(new_dir)
+            try:
+                os.chmod(new_dir, 0o000)
+            except OSError:
+                self.skipTest("platform does not allow "
+                              "changing mode of the cwd")
+
+            # Do not want PermissionError raised.
+            self.assertIsNone(self.machinery.PathFinder.find_spec('whatever'))
+
```

Comment:
If accepting the `change_chmod` context:

```suggestion
    def test_permission_error_cwd(self):
        # gh-115911
        with (
            os_helper.temp_dir() as new_dir,
            os_helper.change_mode(self, new_dir), 0o000),
            os_helper.save_cwd(),
            util.import_state(path=['']),
        ):
            os.chdir(new_dir)

            # Do not want PermissionError raised.
            self.assertIsNone(self.machinery.PathFinder.find_spec('whatever'))

```

Reply:
My initial design was as you suggest, but I wasn't able to make that work across all of the CI platforms and I've concluded it can't be made to work. I found that to avoid any undesired `PermissionError` during setup/teardown
1. during setup `chmod()` and `chdir()` must be called in a a precise order
1. during teardown they require another precise order
1. teardown order cannot be the reverse of setup.

Having the setup `chdir()` & setup `chmod()` calls within the context manager stack makes it impossible to satisfy item 3. Python will always unroll the context manager stack in reverse order. That is why I arrived at `save_cwd()` and `save_mode()` - so the context managers can teardown in the correct order, without imposing the wrong setup order (or vice versa).

As a specific example on macOS the order of operations must be
1. `cd testdir; chmod 000 testdir`;
2. perform test
3. `cd <saved>; chmod <saved> testdir;`

Doing otherwise during setup (chmod, cd) or teardown (chmod, cd) results in permission denied

## Wrong setup order
```
➜  tmp mkdir test1
➜  tmp chmod 000 test1
➜  tmp cd test1
cd: permission denied: test1
```
## Wrong teardown order
```
➜  tmp mkdir test2
➜  tmp cd test2
➜  test2 chmod 000 .
➜  test2 chmod u=rwx .
chmod: .: Permission denied
```
## Correct orders
```
➜  tmp mkdir test3
➜  tmp cd test3
➜  test3 chmod 000 .
➜  test3 cd ..
➜  tmp chmod u=rwx test3
➜  tmp rmdir test3 
```
```
➜  tmp sw_vers
ProductName:		macOS
ProductVersion:		14.6.1
BuildVersion:		23G93
```

Reply:
If you're satisfied with my explanation/example then I could incorporate a condensed version into the doc strings of the context managers and/or add comments in the test case.

Reply:
Wow. That's really annoying. Thanks for tracking it all down and explaining it here. You may want to consider, in addition to a summarized or condensed explanation in the code, also link to the comment with the elaborate explanation. I'm satisfied regardless.

Reply:
Thank you, and will do.

Reply:
> You may want to consider, in addition to a summarized or condensed explanation in the code, also link to the comment with the elaborate explanation.

Done

> I'm satisfied regardless.

Thank you, I'll mark these conversations as resolved.